### PR TITLE
Use default format to fill SPI_SHADER_COL_FORMAT

### PR DESCRIPTION
**Pull request by tywuAMD**
Failures observed when running CTS cases 
dEQP-_VK.shader_object.rendering.color_attachment_count_4.none.extra_output_before_2.\*.random_color_formats.\*_

Those cases randomly insert dummy attachment with format **VK_FORMAT_UNDEFINED** into the valid attachment set.

According to the hardware spec on **SPI_SHADER_COL_FORMAT**, it cannot have holes i.e. if COL2_EXPORT_FORMAT is non-zero, COL0_EXPORT_FORMAT and COL1_EXPORT_FORMAT must also be non-zero.

Thus, use a default value **EXP_FORMAT_FP16_ABGR** instead of **EXP_FORMAT_ZERO** to avoid the above issue. Not quite sure if we need extra check here, thus any suggestion would be appreciated.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit d02c6419a841b164b90c7d52a930e5fd00049aae
<details><summary>CTS tests (Failed: 0/142046)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71028 (56.9%)</li>
<li>Failed: 0/71028 (0.0%)</li>
<li>Not Supported: 30602/71028 (43.1%)</li><li>Warnings: 0/71028 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71018 (57.0%)</li>
<li>Failed: 0/71018 (0.0%)</li>
<li>Not Supported: 30554/71018 (43.0%)</li><li>Warnings: 0/71018 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 69f2fc8585e501c7cc514c9dc3c6e1283e288980
<details><summary>CTS tests (Failed: 1/142046)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71028 (56.9%)</li>
<li>Failed: 0/71028 (0.0%)</li>
<li>Not Supported: 30602/71028 (43.1%)</li><li>Warnings: 0/71028 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40463/71018 (57.0%)</li>
<li><details><summary>Failed: 1/71018 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.glsl.shader_clock.compute.clock2x32ARB
Stack trace: Flake
```
</p></details></li>

<li>Not Supported: 30554/71018 (43.0%)</li><li>Warnings: 0/71018 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 69f2fc8585e501c7cc514c9dc3c6e1283e288980
<details><summary>CTS tests (Failed: 0/142046)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71028 (56.9%)</li>
<li>Failed: 0/71028 (0.0%)</li>
<li>Not Supported: 30602/71028 (43.1%)</li><li>Warnings: 0/71028 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71018 (57.0%)</li>
<li>Failed: 0/71018 (0.0%)</li>
<li>Not Supported: 30554/71018 (43.0%)</li><li>Warnings: 0/71018 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 8f2ba2603c9431fde226015a6f142ecdc512c5bd
<details><summary>CTS tests (Failed: 0/142046)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71028 (56.9%)</li>
<li>Failed: 0/71028 (0.0%)</li>
<li>Not Supported: 30602/71028 (43.1%)</li><li>Warnings: 0/71028 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71018 (57.0%)</li>
<li>Failed: 0/71018 (0.0%)</li>
<li>Not Supported: 30554/71018 (43.0%)</li><li>Warnings: 0/71018 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

**Review by nhaehnle — COMMENTED** [review]
Please check the failures.

There is an issue with this solution, which is that the export format is also used to guide other things in the compiler. For example, it is used in the naming of fragment color glue shaders, and it affects `FragColorExport::handleColorExportInstruction`.

Since the issue is specific to the programming of hardware registers, how about fixing it as part of the config builder? It looks like `PalMetadata::updateSpiShaderColFormat` would be to do the change.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 8f2ba2603c9431fde226015a6f142ecdc512c5bd
<details><summary>CTS tests (Failed: 0/142032)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71014 (56.9%)</li>
<li>Failed: 0/71014 (0.0%)</li>
<li>Not Supported: 30588/71014 (43.1%)</li><li>Warnings: 0/71014 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71018 (57.0%)</li>
<li>Failed: 0/71018 (0.0%)</li>
<li>Not Supported: 30554/71018 (43.0%)</li><li>Warnings: 0/71018 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

**Comment by tywuAMD** [comment]
> Please check the failures.
> 
> There is an issue with this solution, which is that the export format is also used to guide other things in the compiler. For example, it is used in the naming of fragment color glue shaders, and it affects `FragColorExport::handleColorExportInstruction`.
> 
> Since the issue is specific to the programming of hardware registers, how about fixing it as part of the config builder? It looks like `PalMetadata::updateSpiShaderColFormat` would be to do the change.

Thank you so much for the hint. I've uploaded a new patch according to your comment.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1353ebd6d0f7732663f0943f574b6845c6a1637c
<details><summary>CTS tests (Failed: 0/142046)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71028 (56.9%)</li>
<li>Failed: 0/71028 (0.0%)</li>
<li>Not Supported: 30602/71028 (43.1%)</li><li>Warnings: 0/71028 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71018 (57.0%)</li>
<li>Failed: 0/71018 (0.0%)</li>
<li>Not Supported: 30554/71018 (43.0%)</li><li>Warnings: 0/71018 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

**Review by nhaehnle — DISMISSED** [review]
Thanks, LGTM

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1353ebd6d0f7732663f0943f574b6845c6a1637c
<details><summary>CTS tests (Failed: 0/142046)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71028 (56.9%)</li>
<li>Failed: 0/71028 (0.0%)</li>
<li>Not Supported: 30602/71028 (43.1%)</li><li>Warnings: 0/71028 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71018 (57.0%)</li>
<li>Failed: 0/71018 (0.0%)</li>
<li>Not Supported: 30554/71018 (43.0%)</li><li>Warnings: 0/71018 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1353ebd6d0f7732663f0943f574b6845c6a1637c
<details><summary>CTS tests (Failed: 0/142046)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71028 (56.9%)</li>
<li>Failed: 0/71028 (0.0%)</li>
<li>Not Supported: 30602/71028 (43.1%)</li><li>Warnings: 0/71028 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71018 (57.0%)</li>
<li>Failed: 0/71018 (0.0%)</li>
<li>Not Supported: 30554/71018 (43.0%)</li><li>Warnings: 0/71018 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by tywuAMD** [comment]
Sorry I was onto something else for the past few day. After a bit investigation, keeping fragment export shaders unchanged would lead to a mismatch of the colour targets, so that those CTS cases would still fail. I am wondering if it is reasonable to modify the CI tests based on this pull request?

**Comment by amdvlk-admin** [comment]
### Test summary for commit 4af928d5a70fc48bbb1fa4638e63dc5d1d4e9f63
<details><summary>CTS tests (Failed: 0/141320)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40065/70667 (56.7%)</li>
<li>Failed: 0/70667 (0.0%)</li>
<li>Not Supported: 30602/70667 (43.3%)</li><li>Warnings: 0/70667 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40106/70653 (56.8%)</li>
<li>Failed: 0/70653 (0.0%)</li>
<li>Not Supported: 30547/70653 (43.2%)</li><li>Warnings: 0/70653 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
4af928d5a70fc48bbb1fa4638e63dc5d1d4e9f63 Jenkins build error.
CMake Error at /jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/CMakeLists.txt:3 (cmake_minimum_required):
  CMake 3.20.0 or higher is required.  You are running version 3.17.3

**Comment by nhaehnle** [comment]
The failing relocatable shaders test looks suspicious indeed (it seems like it ought to have an export for color), but I worry that this change "fixes" that for the wrong reason.

Do you have a .pipe file for the failing CTS case you were looking at?

**Comment by amdvlk-admin** [comment]
### Test summary for commit 4af928d5a70fc48bbb1fa4638e63dc5d1d4e9f63
<details><summary>CTS tests (Failed: 0/141320)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40065/70667 (56.7%)</li>
<li>Failed: 0/70667 (0.0%)</li>
<li>Not Supported: 30602/70667 (43.3%)</li><li>Warnings: 0/70667 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40106/70653 (56.8%)</li>
<li>Failed: 0/70653 (0.0%)</li>
<li>Not Supported: 30547/70653 (43.2%)</li><li>Warnings: 0/70653 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
9ea7aa2728c151dfc57db55b697ea65e0ef502e3 Jenkins build error.
CMake Error at CMakeLists.txt:47 (include):
  include could not find requested file:

    AmdCmakeHelper


CMake Error at CMakeLists.txt:49 (include):
  include could not find requested file:

    XglSetupAmdGlobalRoots


CMake Error at CMakeLists.txt:50 (include):
  include could not find requested file:

    XglOptions


CMake Error at CMakeLists.txt:51 (include):
  include could not find requested file:

    XglOverrides


CMake Error at CMakeLists.txt:52 (include):
  include could not find requested file:

    XglCompileDefinitions


CMake Error at CMakeLists.txt:53 (include):
  include could not find requested file:

    XglHelper


CMake Error at CMakeLists.txt:55 (include):
  include could not find requested file:

    XglPackaging


CMake Error at CMakeLists.txt:59 (xgl_options):
  Unknown CMake command  xgl_options .

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0362833a82def7ae43bd0b9c41aa6f59fe92a460
<details><summary>CTS tests (Failed: 0/141308)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40067/70655 (56.7%)</li>
<li>Failed: 0/70655 (0.0%)</li>
<li>Not Supported: 30588/70655 (43.3%)</li><li>Warnings: 0/70655 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40106/70653 (56.8%)</li>
<li>Failed: 0/70653 (0.0%)</li>
<li>Not Supported: 30547/70653 (43.2%)</li><li>Warnings: 0/70653 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 76e741855755ba92050e6f990c885f5dccdc4af7
<details><summary>CTS tests (Failed: 0/141308)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40067/70655 (56.7%)</li>
<li>Failed: 0/70655 (0.0%)</li>
<li>Not Supported: 30588/70655 (43.3%)</li><li>Warnings: 0/70655 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40106/70653 (56.8%)</li>
<li>Failed: 0/70653 (0.0%)</li>
<li>Not Supported: 30547/70653 (43.2%)</li><li>Warnings: 0/70653 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
15a8644260b111a51663328257e633a15d3b3d02 Jenkins build error.
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/state/PalMetadata.cpp:903:12: error: unused variable ‘cbShaderMask’ [-Werror=unused-variable]
   unsigned cbShaderMask = 0;
            ^~~~~~~~~~~~
cc1plus: some warnings being treated as errors
[2940/3073] Building CXX object compiler/llpc/llvm/tools/lgc/CMakeFiles/LLVMlgc.dir/patch/PatchResourceCollect.cpp.o
[2941/3073] Building CXX object compiler/llpc/llvm/tools/lgc/disassembler/CMakeFiles/LLVMlgcdis.dir/PalMetadataRegs.cpp.o

**Comment by amdvlk-admin** [comment]
e5c3f879bf87e4e82de9f572fb65e6e6e44ec13e Jenkins build error.
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/state/PalMetadata.cpp:903:12: error: unused variable ‘cbShaderMask’ [-Werror=unused-variable]
   unsigned cbShaderMask = 0;
            ^~~~~~~~~~~~
cc1plus: some warnings being treated as errors
[2943/3073] Building CXX object compiler/llpc/llvm/tools/lgc/CMakeFiles/LLVMlgc.dir/util/Internal.cpp.o
[2944/3073] Building CXX object compiler/llpc/llvm/tools/lgc/CMakeFiles/LLVMlgc.dir/patch/PatchResourceCollect.cpp.o

**Review by nhaehnle — COMMENTED** [review]
Thanks. Now I don't follow what the "undefined targets" stuff is doing or how it can be correct.

`hwColorTarget` is based on the `Location` in SPIR-V, which is supposed to line up with where the color buffers are assigned. So in a situation of API-level color buffers like:

* Buffer 0 bound
* Buffer 1 not bound
* Buffer 2 bound

... I would expect exports to `mrt0` and `mrt2`. Why don't we do it like that?

It may also be a good idea to add a lit test for the behavior here.

**Comment by linqun** [comment]
> Thanks. Now I don't follow what the "undefined targets" stuff is doing or how it can be correct.
> 
> `hwColorTarget` is based on the `Location` in SPIR-V, which is supposed to line up with where the color buffers are assigned. So in a situation of API-level color buffers like:
> 
> * Buffer 0 bound
> * Buffer 1 not bound
> * Buffer 2 bound
> 
> ... I would expect exports to `mrt0` and `mrt2`. Why don't we do it like that?
> 
> It may also be a good idea to add a lit test for the behavior here.

I can't remember which document mentioned this. If I remember correctly, our HW requires export mrt must be continuous. and pack disabled color buffer in cb_shader_mask automatically.
in your previous example, we should program it as following:
cb_shader_mask       = 0x0F0F
spiShaderColFormat = 0x00??
in shader 
exp mrt0 ....
exp mrt1 ....

we follow this rule in both llpc and our internal close source compiler.

**Comment by nhaehnle** [comment]
Thank you for the explanations. I still felt that the solution was a bit complex and hard to follow, so I spent some time to dig around leading to #2546. This is a modification of your change -- can you please check whether it fixes the problems you were seeing?

It is also probably not the last word on this. For example, the logic around CB_SHADER_MASK seems needlessly complicated and I wonder whether we might not be able to build it from `generateExportInstructions` in a similar way as SPI_SHADER_COL_FORMAT. However, there are more open questions around that and I wanted to not change too much at once. I've opened an internal ticket for a related issue.
